### DISCUSSION TITLE
How to use waitUntil with trpc

### DISCUSSION THREAD

**Discussion by Ryanjso** [trpc/trpc#5747] [Q&A] [answered]
Is there a recommended way to use Vercel's new `waitUntil` feature with tRPC?

https://vercel.com/docs/functions/functions-api-reference#waituntil

This isn't working in production, it looks like the lambda is shutting down before the asyncFunctionCall can finish

```
procedure.query(() => {
  waitUntil(asyncFunctionCall())
  return {someData} 
})
```

**Comment by hector** [comment]
Hi, any news on this?

Now they are recommending to use `after` in Next.js v15, is this supported in tRPC?
https://nextjs.org/docs/app/api-reference/functions/after

**Comment by konnsim** [comment]
would love to get an idea on if this is possible, it would simplify things a lot to not need to reach for a background job scheduler (trigger.dev/inngest/etc) for simple things that shouldn't block the response.

**Comment by RiccardoRomagnoli** [comment]
Any up?

**Comment by KATT** [comment]
I don't see why [...]

```ts
const appRouter = router({
  hello: publicProcedure.query(() => {
    waitUntil(asyncFunctionCall())

    return {
      message: 'hello world',
    };
  }),
});
```

... wouldn't work.

Can someone do a minimal reproduction and open a feature request in issues?

**Reply by KATT** [reply]
Also - does it work locally? Fails only on Vercel?

**Reply by KATT** [reply]
It might be that you need to use the [Fetch adapter](https://trpc.io/docs/server/adapters/fetch)

**Reply by juliusmarminge** [reply]
also wondering why it would't _just work_...

**Reply by juliusmarminge** [reply]
![CleanShot 2025-04-07 at 14 51 15](https://github.com/user-attachments/assets/35837679-8e95-4d67-99a5-f5bc157f16d3)

this works just fine locally for me. repo: https://github.com/juliusmarminge/trpc-next_after

**Reply by Ryanjso** [reply]
I'm not using tRPC at the moment, but IIRC it worked locally but not when hosted on Vercel. Like I said in OP the lambda would shut down before waiting. However that was a long time ago, could have changed since then.

**Reply by franckdsf** [reply]
> this works just fine locally for me. repo: https://github.com/juliusmarminge/trpc-next_after

The local server is not serverless, so things like calling a ``promise`` without ``await`` will work because the server remains open. 
However, in a serverless environment, this will fail.

Regarding ``after`` or ``waitUntil``, I haven't had a chance to try them yet—can someone confirm whether they work properly in a serverless environment?

**Reply by juliusmarminge** [reply]
> However, in a serverless environment, this will fail.

the OP asked how to use waitUntil (after is the next.js implementation) with trpc. As far as I can tell, if it works locally, there's nothing wrong with tRPC. We have no control over the runtime so if the server shuts down it's on the deployment provider.

**Comment by RiccardoRomagnoli** [comment, marked as answer]
I used it and it works just fine, I think this can be closed

**Reply by franckdsf** [reply]
@RiccardoRomagnoli in serverless environment like vercel ?

**Reply by RiccardoRomagnoli** [reply]
yes, I'm on Vercel

**Reply by Ryanjso** [reply]
Let me give it a try quick myself, will close when I confirm

**Reply by Ryanjso** [reply]
> I used it and it works just fine, I think this can be closed

Confirming this worked for me as well. Will close thread.

```
    dependencies:
      '@prisma/client':
        specifier: ^5.12.1
        version: 5.22.0(prisma@5.22.0)
      '@tanstack/react-query':
        specifier: ^5.67.1
        version: 5.74.3(react@19.1.0)
      '@trpc/client':
        specifier: npm:@trpc/client
        version: 11.1.0(@trpc/server@11.1.0(typescript@5.8.3))(typescript@5.8.3)
      '@trpc/next':
        specifier: npm:@trpc/next
        version: 11.1.0(@tanstack/react-query@5.74.3(react@19.1.0))(@trpc/client@11.1.0(@trpc/server@11.1.0(typescript@5.8.3))(typescript@5.8.3))(@trpc/react-query@11.1.0(@tanstack/react-query@5.74.3(react@19.1.0))(@trpc/client@11.1.0(@trpc/server@11.1.0(typescript@5.8.3))(typescript@5.8.3))(@trpc/server@11.1.0(typescript@5.8.3))(react-dom@19.1.0(react@19.1.0))(react@19.1.0)(typescript@5.8.3))(@trpc/server@11.1.0(typescript@5.8.3))(next@15.3.0(@playwright/test@1.51.1)(react-dom@19.1.0(react@19.1.0))(react@19.1.0))(react-dom@19.1.0(react@19.1.0))(react@19.1.0)(typescript@5.8.3)
      '@trpc/react-query':
        specifier: npm:@trpc/react-query
        version: 11.1.0(@tanstack/react-query@5.74.3(react@19.1.0))(@trpc/client@11.1.0(@trpc/server@11.1.0(typescript@5.8.3))(typescript@5.8.3))(@trpc/server@11.1.0(typescript@5.8.3))(react-dom@19.1.0(react@19.1.0))(react@19.1.0)(typescript@5.8.3)
      '@trpc/server':
        specifier: npm:@trpc/server
        version: 11.1.0(typescript@5.8.3)
      '@vercel/functions':
        specifier: ^2.0.0
        version: 2.0.0
```
